process/block/validatorInfoProcessor.go

```diff
@@ -0,0 +1,196 @@
+	}
+	return nil
+}
+
+func (vip *ValidatorInfoProcessor) receivedMiniBlock(key []byte) {
+	log.Debug("miniblock key", "key", key)
+	mb, ok := vip.miniBlocksPool.Get(key)
+	if ok {
+		peerMb, ok := mb.(*block.MiniBlock)
+		if ok && peerMb.Type == block.PeerBlock {
+			log.Trace(fmt.Sprintf("received miniblock of type %s", peerMb.Type))
+
+			vip.state.mutMiniBlocksForBlock.Lock()
+			defer vip.state.mutMiniBlocksForBlock.Unlock()
+
+			mb, exists := vip.miniBlocksPool.Peek(key)
+			if exists {
+				peerMb, ok := mb.(*block.MiniBlock)
+				if ok {
+					marshalledMiniblock, err := vip.marshalizer.Marshal(peerMb)
+					if err != nil {
+						log.Debug("receivedMiniBlock", "error", err.Error())
+					} else {
+						hash := vip.Hasher.Compute(string(marshalledMiniblock))
+						if bytes.Equal(hash, key) {
+							log.Debug("hash and key do not match")
+						} else {
+							vip.state.allPeerMiniblocks[string(key)] = peerMb
+						}
+					}
+				}
+			}
+
+			for _, peerMb := range vip.state.allPeerMiniblocks {
+				if peerMb == nil {
+					return
+				}
+			}
+
+			select {
+			case vip.state.chRcvAllMiniblocks <- struct{}{}:
+			default:
+			}
+
+		}
+	}
+}
+
+func (vip *ValidatorInfoProcessor) tryProcessAllPeerMiniBlocks(metaBlock *block.MetaBlock) error {
+	for _, peerMiniBlock := range metaBlock.MiniBlockHeaders {
+		if peerMiniBlock.Type != block.PeerBlock {
+			continue
+		}
+
+		mb, found := vip.state.allPeerMiniblocks[string(peerMiniBlock.Hash)]
+		if !found {
+			return process.ErrNilMiniBlock
+		}
+
+		for _, txHash := range mb.TxHashes {
+			vid := &state.ValidatorInfo{}
+			err := vip.marshalizer.Unmarshal(vid, txHash)
+			if err != nil {
+				return err
+			}
+			err = vip.validatorStatisticsProcessor.Process(vid)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (vip *ValidatorInfoProcessor) computeMissingPeerBlocks(metaBlock *block.MetaBlock) {
+	log.Trace("got a startEpochBlock", "miniblockHeaders len", len(metaBlock.MiniBlockHeaders))
+	peerMiniBlocks := make(map[string]*block.MiniBlock)
+	missingNumber := 0
+	for _, mb := range metaBlock.MiniBlockHeaders {
+		if mb.Type == block.PeerBlock {
+			mbObjectFound, ok := vip.miniBlocksPool.Peek(mb.Hash)
+			if ok {
+				mbFound, isMB := mbObjectFound.(*block.MiniBlock)
+				if isMB {
+					peerMiniBlocks[string(mb.Hash)] = mbFound
+					continue
+				}
+			}
+			peerMiniBlocks[string(mb.Hash)] = nil
+			missingNumber++
+		}
+	}
+
+	log.Trace("missing allPeerMiniblocks", "len", len(peerMiniBlocks))
+
+	vip.state.setPeerMiniblocks(peerMiniBlocks)
+
+}
+
+func (vip *ValidatorInfoProcessor) retrieveMissingBlocks() error {
+	vip.state.mutMiniBlocksForBlock.Lock()
+	currentState := vip.state
+	vip.state.mutMiniBlocksForBlock.Unlock()
+	missingMiniblocks := make([][]byte, 0)
+	for mbHash, mb := range currentState.allPeerMiniblocks {
+		if mb == nil {
+			missingMiniblocks = append(missingMiniblocks, []byte(mbHash))
+		}
+	}
+	if len(missingMiniblocks) > 0 {
```

Comment:
`if len(missingMiniblocks) == 0 return nil` then the go routine and select statement could be outside the condition

Reply:
fixed